sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala

```diff
@@ -464,30 +485,43 @@ object RebaseDateTime {
   final val lastSwitchJulianTs: Long = getLastSwitchTs(julianGregRebaseMap)
 
   /**
-   * An optimized version of [[rebaseJulianToGregorianMicros(ZoneId, Long)]]. This method leverages
-   * the pre-calculated rebasing maps to save calculation. If the rebasing map doesn't contain
-   * information about the current JVM system time zone or `micros` is related to Before Common Era,
-   * the function falls back to the regular unoptimized version.
-   *
-   * Note: The function assumes that the input micros was derived from a local timestamp
-   *       at the default system JVM time zone in the Julian calendar.
+   * An optimized version of [[rebaseJulianToGregorianMicros(TimeZone, Long)]]. This method
+   * leverages the pre-calculated rebasing maps to save calculation. If the rebasing map doesn't
+   * contain information about the given time zone `timeZoneId` or `micros` is related to Before
+   * Common Era, the function falls back to the regular unoptimized version.
    *
+   * @param timeZoneId A string identifier of a time zone.
    * @param micros The number of microseconds since the epoch '1970-01-01T00:00:00Z'
    *               in the Julian calendar. It can be negative.
    * @return The rebased microseconds since the epoch in Proleptic Gregorian calendar.
    */
-  def rebaseJulianToGregorianMicros(micros: Long): Long = {
+  def rebaseJulianToGregorianMicros(timeZoneId: String, micros: Long): Long = {
     if (micros >= lastSwitchJulianTs) {
       micros
     } else {
-      val timeZone = TimeZone.getDefault
-      val tzId = timeZone.getID
-      val rebaseRecord = julianGregRebaseMap.getOrNull(tzId)
+      val rebaseRecord = julianGregRebaseMap.getOrNull(timeZoneId)
       if (rebaseRecord == null || micros < rebaseRecord.switches(0)) {
-        rebaseJulianToGregorianMicros(timeZone, micros)
+        rebaseJulianToGregorianMicros(TimeZone.getTimeZone(timeZoneId), micros)
       } else {
         rebaseMicros(rebaseRecord, micros)
       }
     }
   }
+
+  /**
+   * An optimized version of [[rebaseJulianToGregorianMicros(TimeZone, Long)]]. This method
+   * leverages the pre-calculated rebasing maps to save calculation. If the rebasing map doesn't
+   * contain information about the current JVM system time zone or `micros` is related to Before
+   * Common Era, the function falls back to the regular unoptimized version.
+   *
+   * Note: The function assumes that the input micros was derived from a local timestamp
+   *       at the default system JVM time zone in the Julian calendar.
+   *
+   * @param micros The number of microseconds since the epoch '1970-01-01T00:00:00Z'
+   *               in the Julian calendar. It can be negative.
+   * @return The rebased microseconds since the epoch in Proleptic Gregorian calendar.
+   */
+  def rebaseJulianToGregorianMicros(micros: Long): Long = {
```

Comment:
ditto

Reply:
Actually, it should be used everywhere except rebasing in Parquet/Avro like timestamp formatting.

Reply:
how many places do we have? we should understand the places that rely on JVM timezone.

Reply:
`rebaseJulianToGregorianMicros` is used from:
1. `fromJavaTimestamp()` which is used (55 times) from
    1. `CatalystTypeConverter.TimestampConverter`
    2. `CatalogColumnStat.fromExternalString()`
    3. `Literal.apply()`
    4. `LegacyFastTimestampFormatter.format()`
    5. `LegacySimpleTimestampFormatter.parse()`
    6. `HiveInspectors`
    7. `HadoopTableReader`
    8. `KafkaRecordToRowConverter.toInternalRowWithoutHeaders` & `toInternalRowWithHeaders`
    9. `JdbcUtils.makeGetter()`
    10. `OrcAtomicColumnVector`
    11. `ParquetFilters.timestampToMicros` <-- NEED TO RE-CHECK THIS ONE
    12. In many tests
2. `LegacyFastTimestampFormatter.parse()` which is used (20 times) from:
    1. `CSVInferSchema`
    2. `UnivocityGenerator`
    3. `UnivocityParser`
    4. `JacksonGenerator`
    5. `JacksonParser`
    6. `JsonInferSchema`

Reply:
`rebaseGregorianToJulianMicros` is used from:
1. `toJavaTimestamp` which is used (33 times) from:
   1. `CatalystTypeConverter.TimestampConverter`
   2. `Iso8601TimestampFormatter.format()`
   3. `LegacySimpleTimestampFormatter.format`
   4. `HiveInspectors`
   5. `BaseScriptTransformationExec.outputFieldWriters`
   6. `JdbcUtils.makeSetter`
   7. `OrcFilters.castLiteralValue`
   8. `OrcSerializer.newConverter` 
2. `LegacyFastTimestampFormatter.format()` (the same as for `parse()` above).

Reply:
Then we are fine:
1. when converting between java.sql.Timestamp, we must rely on JVM timezone
2. when reading/writing JSON/CSV with legacy mode, we need to follow the legacy behavior and rebase the timestamp. We should have used the session timezone to rebase, but the ship is already sailed and we have to keep using JVM timezone now.

Reply:
> ParquetFilters.timestampToMicros <-- NEED TO RE-CHECK THIS ONE

This is fine too since the rebasing happens locally on the same JVM. Also users can enable Java 8 for filters, see https://github.com/apache/spark/pull/28696